File: backend/modules/shared/grade_scope.py
"""The MVP curriculum scope: DepEd Grade 6 mathematics, and nothing else.

Shared rather than owned by one feature: the directory decides which grade a
section may belong to, and enrolment decides which grade a learner may join.
Both answers have to be the same one, so the rule is stated here once.

MathSmart teaches one grade. Left to the API alone a Teacher/Administrator
could add Grade 3, hang sections off it, and enrol learners into a curriculum
that has no competencies, no modules and no assessments behind it. The refusal
lives here rather than in each route so the rule is stated once, and so the
frontend and the API cannot drift into disagreeing about what it is.

This is a product-scope rule, not a permission and not a grading decision. It
refuses a request; it never decides a learner's result.
"""

from __future__ import annotations

import re

#: The only grade level MathSmart supports.
MVP_GRADE_LEVEL = 6

#: Numbers in a grade name that could plausibly name a grade level. A year such
#: as 2026 is not one of them, so "Grade 6 (2026)" reads as Grade 6.
_GRADE_LIKE = range(1, 13)

_NUMBERS = re.compile(r"\d+")

SCOPE_REFUSAL = (
    f"MathSmart supports the Grade {MVP_GRADE_LEVEL} curriculum only. "
    f"Grade {MVP_GRADE_LEVEL} is the single grade level, and every section belongs to it."
)

SCOPE_CODE = "grade_scope"


def is_mvp_level(level: int | None) -> bool:
    """Whether this level is the one grade MathSmart teaches."""
    return level == MVP_GRADE_LEVEL


def name_contradicts_level(name: str | None, level: int | None) -> bool:
    """Whether a grade's name claims a different grade level than its number.

    A name carrying no grade-like number claims nothing, so it cannot
    contradict anything. A name carrying one has to agree with the level, which
    is what stops a record reading "Grade 3" while it is stored as level 6.
    """
    if not name or level is None:
        return False

    claimed = {int(found) for found in _NUMBERS.findall(name) if int(found) in _GRADE_LIKE}
    if not claimed:
        return False
    return claimed != {level}

File: src/modules/teacher-admin/grades-sections/utils/grade-scope.js
/**
 * The MVP curriculum scope, as the directory sees it.
 *
 * MathSmart teaches DepEd Grade 6 mathematics and nothing else, so the
 * workspace manages one grade level and the sections under it. The API refuses
 * anything else; this module is how the interface stops offering it in the
 * first place, and how it accounts for records that predate the rule instead
 * of pretending they are not there.
 *
 * The same rule is stated on the server in `backend/modules/teacher_admin/
 * grade_scope.py`. Keep the two in step.
 */

/** The only grade level MathSmart supports. */
export const MVP_GRADE_LEVEL = 6;

/** The fixed name used when the Grade 6 record has not been read yet. */
export const MVP_GRADE_NAME = `Grade ${MVP_GRADE_LEVEL}`;

/**
 * Numbers in a grade name that could plausibly name a grade level. A year such
 * as 2026 is not one of them, so "Grade 6 (2026)" reads as Grade 6.
 */
const GRADE_LIKE = new Set(Array.from({ length: 12 }, (_, index) => index + 1));

/** Whether this record is the one grade MathSmart teaches. */
export function isMvpGrade(grade) {
  return Number(grade?.level) === MVP_GRADE_LEVEL;
}

/**
 * Whether a grade's name claims a different grade level than its number.
 *
 * A name carrying no grade-like number claims nothing, so it cannot contradict
 * anything. A name carrying one has to agree with the level, which is what
 * stops a record reading "Grade 3" while it is stored as Grade 6.
 */
export function nameContradictsLevel(name, level) {
  if (typeof name !== "string" || !name.trim()) return false;
  if (level === null || level === undefined) return false;

  const claimed = new Set(
    (name.match(/\d+/g) ?? []).map(Number).filter((found) => GRADE_LIKE.has(found)),
  );
  if (claimed.size === 0) return false;
  return claimed.size !== 1 || !claimed.has(Number(level));
}

/**
 * Splits the directory into what this workspace manages and what it does not.
 *
 * Everything the API returns is accounted for. The Grade 6 record and its
 * sections are the working directory; a legacy grade at another level, a
 * duplicate Grade 6, and any section hanging off one of them are set aside so
 * a teacher can retire them rather than lose sight of them.
 *
 * @param {{grades?: Array, sections?: Array}} directory
 */
export function partitionDirectory({ grades, sections } = {}) {
  const allGrades = Array.isArray(grades) ? grades : [];
  const allSections = Array.isArray(sections) ? sections : [];

  const supported = allGrades.filter(isMvpGrade);
  const grade = supported[0] ?? null;

  // A second Grade 6 row is out of scope too: the product has one grade, and
  // leaving a duplicate in the working list would make sections ambiguous.
  const outOfScopeGrades = allGrades.filter((entry) => entry !== grade);

  const inScope = (section) => grade !== null && section?.grade_id === grade.grade_id;

  return {
    grade,
    outOfScopeGrades,
    sections: allSections.filter(inScope),
    outOfScopeSections: allSections.filter((section) => !inScope(section)),
  };
}

/**
 * The display name for a grade a section points at.
 *
 * Falls back to the raw identifier rather than inventing a name, so a section
 * whose grade is missing reads as a problem instead of as Grade 6.
 */
export function gradeNameFor(section, grades) {
  const match = (Array.isArray(grades) ? grades : []).find(
    (grade) => grade.grade_id === section?.grade_id,
  );
  return match?.name ?? section?.grade_id ?? "Unknown grade";
}
